File: ts/src/chunking.ts
const PHRASAL_VERBS: Record<string, string[]> = {
  set: ["up", "down", "out", "forth", "aside"],
  tear: ["down", "apart", "up"],
  break: ["down", "up", "apart", "out", "into"],
  carry: ["out", "on", "through", "forward"],
  take: ["on", "over", "up", "out", "apart"],
  put: ["together", "aside", "off", "on", "forward"],
  turn: ["on", "off", "up", "down", "into"],
  bring: ["up", "in", "on", "together", "about"],
  come: ["up", "in", "on", "across", "apart"],
  go: ["through", "over", "ahead", "forward", "back"],
  look: ["into", "over", "up", "through", "for"],
  point: ["out", "to", "at"],
  figure: ["out"],
  find: ["out", "a way"],
  run: ["through", "on", "over"],
  move: ["forward", "on", "ahead"],
  back: ["up", "off", "away", "out"],
  call: ["out", "up", "on", "for"],
  map: ["out", "to"],
  scope: ["out", "down"],
  roll: ["out", "up", "back"],
  reach: ["out", "in"],
  follow: ["up", "through", "on"],
  wire: ["up", "together"],
  hook: ["up", "in", "together"],
  build: ["out", "up", "on", "upon"],
  clean: ["up", "out", "off"],
  cut: ["down", "off", "out", "up"],
  phase: ["out", "in"],
  scale: ["up", "down", "out"],
  spin: ["up", "out", "off"],
  stand: ["up", "by"],
  start: ["up", "out", "over"],
  step: ["up", "back", "through", "aside"],
  strip: ["out", "down"],
  sum: ["up"],
  tie: ["together", "in", "into"],
  try: ["out", "on"],
  use: ["up"],
  write: ["out", "up", "down", "off"],
  zero: ["in"],
}

const CLAUSE_BOUNDARIES: string[] = [
  "\\band\\b(?!\\s+then\\b)",
  "\\bbut\\b",
  "\\bor\\b",
  "\\bso that\\b",
  "\\bsuch that\\b",
  "\\bprovided that\\b",
  "\\bhowever\\b",
  "\\btherefore\\b",
  "\\bthen\\b",
  "\\bnext\\b",
  "\\bafterwards\\b",
  "\\bmeanwhile\\b",
  "\\badditionally\\b",
  "\\bfurthermore\\b",
  "\\bmoreover\\b",
  "\\balso\\b",
  "\\bconversely\\b",
  "\\balternatively\\b",
  "\\botherwise\\b",
  "\\bfinally\\b",
  "\\bsubsequently\\b",
  "\\bconsequently\\b",
  "\\bas a result\\b",
]

const TOPIC_SHIFT_MARKERS: string[] = [
  "\\bnow\\b",
  "\\bok[ay]*,?\\s",
  "\\balright,?\\s",
  "\\bso,?\\s",
  "\\bright,?\\s",
  "\\bfirst(ly)?\\b",
  "\\bsecond(ly)?\\b",
  "\\bthird(ly)?\\b",
  "\\bfinally\\b",
  "\\blast(ly)?\\b",
  "\\bin addition\\b",
  "\\bon the other hand\\b",
  "\\bmeanwhile\\b",
  "\\bmoving on\\b",
  "\\bnext up\\b",
]

const CONTRADICTION_MARKERS: RegExp[] = [
  /\b(but|however|although|though|while|whereas|conversely|nevertheless|on the other hand)\b/i,
  /\b(yet|still|instead|rather|otherwise|alternatively)\b/i,
  /\b(must|required|essential)\b.*\b(but|however|although)\b/i,
  /\b(only|exactly|strictly)\b.*\b(also|additionally|furthermore)\b/i,
]

export interface Clause {
  text: string
  verbs: string[]
  constraints: string[]
  has_negation: boolean
  has_hedging: boolean
  has_emphasis: boolean
}

export interface CompoundVerb {
  verb: string
  particle: string
  full: string
  position: number
}

export interface ChunkResult {
  text: string
  clauses: Clause[]
  compound_verbs: CompoundVerb[]
  topic_shifts: number
  contradiction_hits: string[]
  clause_indicators: string[]
}

function escapeRegex(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")
}

function _splitClauses(text: string): string[] {
  const sentences = text.split(/[.!?]+/)
  const result: string[] = []
  const boundaryPattern = "(" + CLAUSE_BOUNDARIES.join("|") + ")"
  const re = new RegExp(boundaryPattern)

  for (let raw of sentences) {
    raw = raw.trim()
    if (!raw) continue
    const parts = raw.split(new RegExp(boundaryPattern))
    let current = ""
    for (const part of parts) {
      if (part && part.trim()) {
        current += part + " "
        if (CLAUSE_BOUNDARIES.some((p) => new RegExp(p).test(part.trim()))) {
          if (current.trim()) {
            result.push(current.trim())
          }
          current = ""
        }
      }
    }
    if (current.trim()) {
      result.push(current.trim())
    }
  }
  return result
}

function _detectCompoundVerbs(tokens: string[]): CompoundVerb[] {
  const found: CompoundVerb[] = []
  for (let i = 0; i < tokens.length; i++) {
    const low = tokens[i].toLowerCase()
    const particles = PHRASAL_VERBS[low]
    if (particles && i + 1 < tokens.length) {
      const nextToken = tokens[i + 1].toLowerCase().replace(/[.!;,]+$/, "")
      if (particles.includes(nextToken)) {
        found.push({
          verb: low,
          particle: nextToken,
          full: `${low}_${nextToken}`,
          position: i,
        })
      }
    }
  }
  return found
}

function _detectContradictions(clauses: Clause[]): string[] {
  const hits = new Set<string>()
  for (let i = 0; i < clauses.length; i++) {
    for (const pattern of CONTRADICTION_MARKERS) {
      if (pattern.test(clauses[i].text)) {
        for (let j = i + 1; j < clauses.length; j++) {
          if (clauses[i].has_negation && clauses[j].has_emphasis) {
            hits.add(`clause ${i + 1} (negation) vs clause ${j + 1} (emphasis)`)
          } else if (clauses[i].has_emphasis && clauses[j].has_negation) {
            hits.add(`clause ${i + 1} (emphasis) vs clause ${j + 1} (negation)`)
          }
        }
      }
    }
  }
  return [...hits]
}

function _countTopicShifts(clauses: Clause[]): number {
  let count = 0
  for (const clause of clauses) {
    for (const marker of TOPIC_SHIFT_MARKERS) {
      if (new RegExp(marker, "i").test(clause.text)) {
        count++
      }
    }
  }
  return count
}

function _detectVerbsInClause(clauseText: string, knownVerbs: Set<string>): string[] {
  const lower = clauseText.toLowerCase()
  const found: string[] = []
  for (const v of knownVerbs) {
    if (new RegExp(`\\b${escapeRegex(v)}\\b`).test(lower)) {
      found.push(v)
    }
  }
  return found
}

function _detectConstraintsInClause(clauseText: string): string[] {
  const lower = clauseText.toLowerCase()
  const constraints: string[] = []
  if (/\b(without|avoid|never|no |not|don't|do not|except|excluding)\b/.test(lower)) {
    constraints.push("negation")
  }
  if (/\b(must|need to|have to|required|essential)\b/.test(lower)) {
    constraints.push("requirement")
  }
  if (/\b(only|exactly|strictly|specifically|precisely)\b/.test(lower)) {
    constraints.push("exact")
  }
  if (/\b(using|with|via|through|by|import|require|dependency|library)\b/.test(lower)) {
    constraints.push("dependency")
  }
  return constraints
}

export function chunk(text: string, knownVerbs?: Set<string>): ChunkResult {
  const tokens = text.match(/\b\w+\b/g) ?? []
  const verbSet = knownVerbs ?? new Set<string>()
  const compound_verbs = _detectCompoundVerbs(tokens)
  const clauseTexts = _splitClauses(text)

  const clauses: Clause[] = []
  for (const ct of clauseTexts) {
    const verbsIn = _detectVerbsInClause(ct, verbSet)
    const constraintsIn = _detectConstraintsInClause(ct)
    const hasNeg = constraintsIn.includes("negation")
    const hasHedge = /\b(maybe|perhaps|possibly|probably|might|could|i think|i believe)\b/i.test(ct)
    const hasEmp = /\b(must|absolutely|critical|urgent|essential|required|vital)\b/i.test(ct)
    clauses.push({
      text: ct.trim(),
      verbs: verbsIn,
      constraints: constraintsIn,
      has_negation: hasNeg,
      has_hedging: hasHedge,
      has_emphasis: hasEmp,
    })
  }

  const contradiction_hits = _detectContradictions(clauses)
  const topic_shifts = _countTopicShifts(clauses)

  const clause_indicators: string[] = []
  if (clauses.length > 3) {
    clause_indicators.push(`${clauses.length} clauses detected (may indicate multi-instruction)`)
  }
  if (contradiction_hits.length > 0) {
    clause_indicators.push(`contradictory markers: ${contradiction_hits.slice(0, 2).join(", ")}`)
  }
  if (compound_verbs.length > 0) {
    clause_indicators.push(`phrasal verbs: ${compound_verbs.slice(0, 3).map((cv) => cv.full).join(", ")}`)
  }
  if (topic_shifts > 2) {
    clause_indicators.push(`${topic_shifts} topic shifts (may indicate scope creep)`)
  }

  return {
    text,
    clauses,
    compound_verbs,
    topic_shifts,
    contradiction_hits,
    clause_indicators,
  }
}

export function renderChunkReport(cr: ChunkResult): string {
  const sep = "=".repeat(56)
  const lines: string[] = [sep, "  Chunk analysis", sep, ""]

  if (cr.compound_verbs.length > 0) {
    lines.push(`  Phrasal verbs: ${cr.compound_verbs.map((cv) => cv.full).join(", ")}`)
    lines.push("")
  }

  lines.push(`  Clauses (${cr.clauses.length}):`)
  for (let i = 0; i < cr.clauses.length; i++) {
    const clause = cr.clauses[i]
    const parts: string[] = []
    if (clause.has_negation) parts.push("NEG")
    if (clause.has_emphasis) parts.push("EMP")
    if (clause.has_hedging) parts.push("HEDGE")
    const tags = parts.length > 0 ? `  [${parts.join(", ")}]` : ""
    const short = clause.text.slice(0, 55)
    lines.push(`    ${i + 1}. ${short}${tags}`)
    if (clause.verbs.length > 0) {
      lines.push(`       verbs: ${clause.verbs.join(", ")}`)
    }
    if (clause.constraints.length > 0) {
      lines.push(`       constraints: ${clause.constraints.join(", ")}`)
    }
  }

  if (cr.contradiction_hits.length > 0) {
    lines.push("")
    lines.push("  Potential contradictions:")
    for (const h of cr.contradiction_hits) {
      lines.push(`    ! ${h}`)
    }
  }

  if (cr.topic_shifts > 2) {
    lines.push("")
    lines.push(`  Topic shifts: ${cr.topic_shifts} (possible scope creep)`)
  }

  lines.push(sep)
  return lines.join("\n")
}

File: src/ambiguity/chunking.py
"""Clause-level chunking and multi-word verb detection — zero-dependency."""

import re
from dataclasses import dataclass, field

# Multi-word phrasal verbs — verb + particle combinations
PHRASAL_VERBS: dict[str, list[str]] = {
    "set": ["up", "down", "out", "forth", "aside"],
    "tear": ["down", "apart", "up"],
    "break": ["down", "up", "apart", "out", "into"],
    "carry": ["out", "on", "through", "forward"],
    "take": ["on", "over", "up", "out", "apart"],
    "put": ["together", "aside", "off", "on", "forward"],
    "turn": ["on", "off", "up", "down", "into"],
    "bring": ["up", "in", "on", "together", "about"],
    "come": ["up", "in", "on", "across", "apart"],
    "go": ["through", "over", "ahead", "forward", "back"],
    "look": ["into", "over", "up", "through", "for"],
    "point": ["out", "to", "at"],
    "figure": ["out"],
    "find": ["out", "a way"],
    "run": ["through", "on", "over"],
    "move": ["forward", "on", "ahead"],
    "back": ["up", "off", "away", "out"],
    "call": ["out", "up", "on", "for"],
    "map": ["out", "to"],
    "scope": ["out", "down"],
    "roll": ["out", "up", "back"],
    "reach": ["out", "in"],
    "follow": ["up", "through", "on"],
    "wire": ["up", "together"],
    "hook": ["up", "in", "together"],
    "build": ["out", "up", "on", "upon"],
    "clean": ["up", "out", "off"],
    "cut": ["down", "off", "out", "up"],
    "phase": ["out", "in"],
    "scale": ["up", "down", "out"],
    "spin": ["up", "out", "off"],
    "stand": ["up", "by"],
    "start": ["up", "out", "over"],
    "step": ["up", "back", "through", "aside"],
    "strip": ["out", "down"],
    "sum": ["up"],
    "tie": ["together", "in", "into"],
    "try": ["out", "on"],
    "use": ["up"],
    "write": ["out", "up", "down", "off"],
    "zero": ["in"],
}

# Clause boundary markers — split on these to separate instructions
CLAUSE_BOUNDARIES: list[str] = [
    r"\band\b(?!\s+then\b)",
    r"\bbut\b",
    r"\bor\b",
    r"\bso that\b",
    r"\bsuch that\b",
    r"\bprovided that\b",
    r"\bhowever\b",
    r"\btherefore\b",
    r"\bthen\b",
    r"\bnext\b",
    r"\bafterwards\b",
    r"\bmeanwhile\b",
    r"\badditionally\b",
    r"\bfurthermore\b",
    r"\bmoreover\b",
    r"\balso\b",
    r"\bconversely\b",
    r"\balternatively\b",
    r"\botherwise\b",
    r"\bfinally\b",
    r"\bsubsequently\b",
    r"\bconsequently\b",
    r"\bas a result\b",
]

# Topic shift markers — stronger boundaries indicating new instructions
TOPIC_SHIFT_MARKERS: list[str] = [
    r"\bnow\b",
    r"\bok[ay]*,?\s",
    r"\balright,?\s",
    r"\bso,?\s",
    r"\bright,?\s",
    r"\bfirst(ly)?\b",
    r"\bsecond(ly)?\b",
    r"\bthird(ly)?\b",
    r"\bfinally\b",
    r"\blast(ly)?\b",
    r"\bin addition\b",
    r"\bon the other hand\b",
    r"\bmeanwhile\b",
    r"\bmoving on\b",
    r"\bnext up\b",
]

# Contradiction markers — signals that two clauses may conflict
CONTRADICTION_MARKERS: list[re.Pattern] = [
    re.compile(r"\b(but|however|although|though|while|whereas|conversely|nevertheless|on the other hand)\b", re.IGNORECASE),
    re.compile(r"\b(yet|still|instead|rather|otherwise|alternatively)\b", re.IGNORECASE),
    re.compile(r"\b(must|required|essential)\b.*\b(but|however|although)\b", re.IGNORECASE),
    re.compile(r"\b(only|exactly|strictly)\b.*\b(also|additionally|furthermore)\b", re.IGNORECASE),
]


@dataclass
class Clause:
    text: str
    verbs: list[str] = field(default_factory=list)
    constraints: list[str] = field(default_factory=list)
    has_negation: bool = False
    has_hedging: bool = False
    has_emphasis: bool = False


@dataclass
class CompoundVerb:
    verb: str
    particle: str
    full: str
    position: int = 0


@dataclass
class ChunkResult:
    text: str
    clauses: list[Clause] = field(default_factory=list)
    compound_verbs: list[CompoundVerb] = field(default_factory=list)
    topic_shifts: int = 0
    contradiction_hits: list[str] = field(default_factory=list)

    @property
    def clause_count(self) -> int:
        return len(self.clauses)

    @property
    def has_contradictions(self) -> bool:
        return len(self.contradiction_hits) > 0

    @property
    def clause_indicators(self) -> list[str]:
        indicators = []
        if self.clause_count > 3:
            indicators.append(f"{self.clause_count} clauses detected (may indicate multi-instruction)")
        if self.has_contradictions:
            indicators.append(f"contradictory markers: {', '.join(self.contradiction_hits[:2])}")
        if self.compound_verbs:
            verbs = [f"{cv.full}" for cv in self.compound_verbs[:3]]
            indicators.append(f"phrasal verbs: {', '.join(verbs)}")
        if self.topic_shifts > 2:
            indicators.append(f"{self.topic_shifts} topic shifts (may indicate scope creep)")
        return indicators


def _split_clauses(text: str) -> list[str]:
    """Split text into clauses using boundary markers."""
    # First split on sentence boundaries
    clauses = re.split(r'[.!?]+', text)
    result = []
    for clause in clauses:
        clause = clause.strip()
        if not clause:
            continue
        # Further split on clause boundaries
        # Build a combined pattern from CLAUSE_BOUNDARIES
        boundary_pattern = "(" + "|".join(CLAUSE_BOUNDARIES) + ")"
        # Split preserving the boundary marker context
        parts = re.split(boundary_pattern, clause)
        current = ""
        for part in parts:
            if part and part.strip():
                current += part + " "
                # If the last word is a boundary marker, flush current
                if any(re.search(p, part.strip()) for p in CLAUSE_BOUNDARIES):
                    if current.strip():
                        result.append(current.strip())
                    current = ""
        if current.strip():
            result.append(current.strip())
    return result


def _detect_compound_verbs(tokens: list[str]) -> list[CompoundVerb]:
    """Detect multi-word phrasal verbs in token stream."""
    found: list[CompoundVerb] = []
    for i, token in enumerate(tokens):
        low = token.lower()
        if low in PHRASAL_VERBS:
            # Check next token for particle
            if i + 1 < len(tokens):
                next_token = tokens[i + 1].lower().rstrip(".!,;")
                if next_token in PHRASAL_VERBS[low]:
                    found.append(CompoundVerb(
                        verb=low,
                        particle=next_token,
                        full=f"{low}_{next_token}",
                        position=i,
                    ))
    return found


def _detect_contradictions(clauses: list[Clause]) -> list[str]:
    """Detect contradictory constraint pairs across clauses."""
    hits: list[str] = []
    for i, clause in enumerate(clauses):
        for pattern in CONTRADICTION_MARKERS:
            if pattern.search(clause.text):
                # Look for contradictory indicators in other clauses
                for j in range(i + 1, len(clauses)):
                    # Check if one clause has negation and another has requirement
                    if clauses[i].has_negation and clauses[j].has_emphasis:
                        hits.append(f"clause {i+1} (negation) vs clause {j+1} (emphasis)")
                    elif clauses[i].has_emphasis and clauses[j].has_negation:
                        hits.append(f"clause {i+1} (emphasis) vs clause {j+1} (negation)")
    return list(set(hits))


def _count_topic_shifts(clauses: list[Clause]) -> int:
    """Count topic shift markers across clauses."""
    count = 0
    for clause in clauses:
        for marker in TOPIC_SHIFT_MARKERS:
            if re.search(marker, clause.text, re.IGNORECASE):
                count += 1
    return count


def _detect_verbs_in_clause(clause_text: str, known_verbs: set[str]) -> list[str]:
    """Find known verbs present in a clause."""
    lower = clause_text.lower()
    return [v for v in known_verbs if re.search(r"\b" + re.escape(v) + r"\b", lower)]


def _detect_constraints_in_clause(clause_text: str) -> list[str]:
    """Detect constraint types present in a clause."""
    lower = clause_text.lower()
    constraints = []
    if re.search(r"\b(without|avoid|never|no |not|don't|do not|except|excluding)\b", lower):
        constraints.append("negation")
    if re.search(r"\b(must|need to|have to|required|essential)\b", lower):
        constraints.append("requirement")
    if re.search(r"\b(only|exactly|strictly|specifically|precisely)\b", lower):
        constraints.append("exact")
    if re.search(r"\b(using|with|via|through|by|import|require|dependency|library)\b", lower):
        constraints.append("dependency")
    return constraints


def chunk(text: str, known_verbs: set[str] | None = None) -> ChunkResult:
    """Analyze text into clauses, compound verbs, and contradiction patterns."""
    tokens = re.findall(r"\b\w+\b", text)
    known_verbs = known_verbs or set()
    verb_set = known_verbs

    # Detect compound verbs
    compound_verbs = _detect_compound_verbs(tokens)

    # Split into clauses
    clause_texts = _split_clauses(text)

    # Analyze each clause
    clauses: list[Clause] = []
    for ct in clause_texts:
        verbs_in = _detect_verbs_in_clause(ct, verb_set)
        constraints_in = _detect_constraints_in_clause(ct)
        has_neg = "negation" in constraints_in
        has_hedge = bool(re.search(
            r"\b(maybe|perhaps|possibly|probably|might|could|i think|i believe)\b",
            ct, re.IGNORECASE
        ))
        has_emp = bool(re.search(
            r"\b(must|absolutely|critical|urgent|essential|required|vital)\b",
            ct, re.IGNORECASE
        ))
        clauses.append(Clause(
            text=ct.strip(),
            verbs=verbs_in,
            constraints=constraints_in,
            has_negation=has_neg,
            has_hedging=has_hedge,
            has_emphasis=has_emp,
        ))

    # Detect contradictions
    contradiction_hits = _detect_contradictions(clauses)

    # Count topic shifts
    topic_shifts = _count_topic_shifts(clauses)

    return ChunkResult(
        text=text,
        clauses=clauses,
        compound_verbs=compound_verbs,
        topic_shifts=topic_shifts,
        contradiction_hits=contradiction_hits,
    )


def render_chunk_report(cr: ChunkResult) -> str:
    sep = "=" * 56
    lines = [sep]
    lines.append("  Chunk analysis")
    lines.append(sep)
    lines.append("")

    if cr.compound_verbs:
        verbs_str = ", ".join(f"{cv.full}" for cv in cr.compound_verbs)
        lines.append(f"  Phrasal verbs: {verbs_str}")
        lines.append("")

    lines.append(f"  Clauses ({cr.clause_count}):")
    for i, clause in enumerate(cr.clauses, 1):
        parts = []
        if clause.has_negation:
            parts.append("NEG")
        if clause.has_emphasis:
            parts.append("EMP")
        if clause.has_hedging:
            parts.append("HEDGE")
        tags = f"  [{', '.join(parts)}]" if parts else ""
        short = clause.text[:55]
        lines.append(f"    {i}. {short}{tags}")
        if clause.verbs:
            lines.append(f"       verbs: {', '.join(clause.verbs)}")
        if clause.constraints:
            lines.append(f"       constraints: {', '.join(clause.constraints)}")

    if cr.contradiction_hits:
        lines.append("")
        lines.append("  Potential contradictions:")
        for h in cr.contradiction_hits:
            lines.append(f"    ! {h}")

    if cr.topic_shifts > 2:
        lines.append("")
        lines.append(f"  Topic shifts: {cr.topic_shifts} (possible scope creep)")

    lines.append(sep)
    return "\n".join(lines)
